### ISSUE TITLE
Weird error message for incompatible override when using a `TypeVar` for `self`

### ISSUE THREAD

**Issue by AlexWaygood** [python/mypy#12786]
**Bug Report**

Minimal repro:

```python
from typing import TypeVar

Self = TypeVar("Self")

class Flag:
    def __or__(self: Self, other: Self) -> Self: ...
    __ror__ = __or__

class IntFlag(int, Flag):
    def __or__(self: Self, other: int) -> Self: ...
    def __ror__(self: Self, other: int) -> Self: ...
```

mypy output:
```
main.py:11: error: Signature of "__ror__" incompatible with supertype "Flag"
main.py:11: note:      Superclass:
main.py:11: note:          def [Self] __or__(self, Self, Self) -> Self
main.py:11: note:      Subclass:
main.py:11: note:          def __ror__(self, int) -> IntFlag
```

Mypy playground link here: https://mypy-play.net/?mypy=latest&python=3.10&gist=26b60881cd02e608bb9e555db4c3f627

**Expected Behavior**

I'm not sure an error is even appropriate here: no incompatible override is reported for `__or__`.

Even if an error is appropriate here, however, the error message is weird and garbled. According to the error message, mypy seems to think that `__ror__` in the superclass expects three arguments (it only expects two). It also seems to think the method in the subclass will return `IntFlag` (not necessarily true; the return type will be whatever `self` is, and `self` could be any subtype of `IntFlag`).

**Mypy version used**

- 0.950.
- This doesn't seem to be a recent regression: [I can reproduce as far back as 0.920](https://mypy-play.net/?mypy=0.920&python=3.10&gist=3bbdee8e75fe30fe8ebe6ca8a4b50ce8), which is the earliest mypy version that has the verbose error messages.
- I've also reproduced the bug on mypy master.

**Comment by AlexWaygood** [comment]
I think the root cause of the error here is the fact that mypy erroneously believes `__ror__` in the base class to be a callable instance variable, rather than a method:

```python
from typing import TypeVar

Self = TypeVar("Self")

class Flag:
    def __or__(self: Self, other: Self) -> Self: ...
    __ror__ = __or__

reveal_type(Flag.__or__)  # Revealed type is "def [Self] (Self`-1, Self`-1) -> Self`-1"
reveal_type(Flag.__ror__)  # error: Access to generic instance variables via class is ambiguous  # Revealed type is "def [Self] (Any, Any) -> Any"
```

https://mypy-play.net/?mypy=latest&python=3.10&gist=daa1934028156c5298f443fc65c1afd4
